CONTRIBUTING.md

```diff
@@ -0,0 +1,271 @@
+-   After pressing escape key from the insert mode, extension sends changes obtained from the insert mode to Neovim.
+
+## Neovim APIs
+
+### Node Client
+
+VSCode-Neovim uses neovim's [`node-client`](https://neovim.io/node-client/index.html) library to communicate with
+neovim. This library is a wrapper around neovim's [msgpack-rpc API](https://neovim.io/doc/user/api.html). The library is
+used to send and receive messages to and from neovim. The library is also used to spawn neovim processes. Many functions
+are exposed, listed [here](https://neovim.io/node-client/classes/NeovimClient.html). However, the library is not kept up
+to date, so many API wrappers are missing. In these cases, the raw `call`, `callAtomic`, `lua`, and `command` commands
+are directly used.
+
+### Neovim UI Protocol
+
+VSCode-Neovim is built around neovim's ui protocol [:help ui](https://neovim.io/doc/user/ui.html). This protocol allows
+vscode to listen to events emitted by neovim, such as `mode_change`, `hl_attr_define`, and `win_viewport`. This API
+allows neovim to externalize the _rendering_ of various components of the editor, such as the cursor, highlights,
+buffers, cmdline, etc. When possible, this protocol should be used as the source of truth for the state of the editor.
+
+However, the UI protocol [is incomplete](https://github.com/neovim/neovim/issues/9421), and there are many holes in the
+API. For example, there is no way to get the type of visual mode or the current visual selection, `win_viewport` is not
+triggered on horizontal scrolling, there are no events for window splits and tabs, etc. As the nvim UI extension API is
+improved, these holes should be filled in, and workarounds can be removed.
+
+### Autocommands
+
+Another event-driven API available is [autocommands](https://neovim.io/doc/user/autocmd.html). These are events that are
+triggered when certain actions happen, such as `BufEnter`, `BufWrite`, `CursorMoved`, etc. These events populate
+`<afile>`, `vim.v.event`, etc. All this data is available using
+[nvim_create_autocmd](<https://neovim.io/doc/user/api.html#nvim_create_autocmd()>). Sometimes, these events contain more
+information than the UI protocol, and should be used instead. For example, the `ModeChanged` autocommand contains the
+type of visual mode, which is not available in the UI protocol. In that case, the autocommand should forward the event
+to vscode for custom handling.
+
+### Neovim Lua API
+
+To send commands to neovim, and to (worst-case scenario) ask it for additional information, the
+[neovim lua api](https://neovim.io/doc/user/api.html) can be used. As a last resort, the
+[builtin](https://neovim.io/doc/user/builtin.html) API can be used.
+
+Due to the number of RPC round trips required, the logic using the lua or vim APIs should be moved to a custom lua
+function running in nvim.
+
+## Managers
+
+VSCode-Neovim is structured using a manager pattern. Each manager is responsible for syncing a specific aspect of the
```

Comment:
should we mention the eventbus `.on` ? https://github.com/vscode-neovim/vscode-neovim/blob/c554c065a375f2bd32346aabdb6526ece8bb4189/src/eventBus.ts#L176